### PR TITLE
#3934: Enable DPRINT on N300 Device 0

### PR DESCRIPTION
**Pull request by tt-dma**
Changes /fixes to the DPRINT server + tests to enable running on n300 device 0. Multi-device will be left for a separate PR.

Also some general refactoring/cleanup in the print server.

### CODE REVIEW

File: tt_metal/llrt/tt_cluster.cpp [hunk 1]
```diff
@@ -346,30 +346,6 @@ int Cluster::get_device_aiclk(const chip_id_t &chip_id) const {
     return 0;
 }
 
-void Cluster::reset_debug_print_server_buffers() const {
-    for (const auto &[device_id, mmio_device_id] : this->device_to_mmio_device_) {
-        auto workers = get_soc_desc(device_id).workers;
```
Comment (tt-dma) [inline tt_metal/llrt/tt_cluster.cpp:351]: The issue that was causing hangs is from here: the `workers` vector used virtual core coords, but `write_core` requires physical core coords (it virtualizes them further down). This ultimately ended up with writes to harvested cores on n300 in some cases.

We already get a list of all physical cores in the print server, so moving this logic there and not using `workers` makes things work as intended

File: tt_metal/impl/debug/dprint_server.cpp [hunk 2]
```diff
@@ -206,25 +172,174 @@ static void print_tile_slice(ostream& stream, uint8_t* ptr, int hart_id) {
...
+        std::this_thread::sleep_for(std::chrono::milliseconds(5));
+    } while (hart_waiting_on_signal_.size() > 0 || new_data_last_iter_);
+} // WaitForPrintsFinished
+
+void DebugPrintServerContext::AttachDevice(Device* device) {
+    chip_id_t device_id = device->id();
+
+    // A set of all valid worker cores, used for checking the user input. Note that the core coords
+    // here are physical, which matches the user input for configuring the print server.
+    auto compare_coords = [](const CoreCoord& a, const CoreCoord& b){
```
Comment (aliuTT) [inline tt_metal/impl/debug/dprint_server.cpp:264]: Doesn't CoreCoord already have these defined? it's in `tt_xy_pair.h`

File: tt_metal/impl/debug/dprint_server.hpp [hunk 3]
```diff
@@ -22,46 +24,38 @@ enum DebugPrintHartFlags : unsigned int {
 constexpr int DPRINT_NRISCVS = 5;
 
 /*
-@brief Starts the print server thread - will poll all specified chip/cores/harts of a device for any print data accumulated in thread-local buffers.
-
-thread_mask in this API is using flags from DebugPrintHartFlags to enable per-thread debug server listening.
-This call will launch a host thread for each core in cores array and each hart specified by the mask.
-
-Note that this call only works correctly after open_device and start_device calls. (TODO(AP): add runtime checks)
+@brief Attachess a device to be monitored by the print server. If no devices were present on the
```
Comment (aliuTT) [inline tt_metal/impl/debug/dprint_server.hpp:27]: typo, but feel free to ignore lol

**Comment by aliuTT** [comment]
What dprint features are working now? Are you able to enable and see dprint dumped from a remote chip?

File: tt_metal/impl/debug/dprint_server.hpp (hunk 3 above)
Comment (tt-dma) [reply tt_metal/impl/debug/dprint_server.hpp]: Fixed!

File: tt_metal/impl/debug/dprint_server.cpp (hunk 2 above)
Comment (tt-dma) [reply tt_metal/impl/debug/dprint_server.cpp]: Missed this yeah, removed this redundant function in latest commit

**Comment by tt-dma** [comment]
> What dprint features are working now? Are you able to enable and see dprint dumped from a remote chip?

Currently all dprint features/tests (fast and slow dispatch) are working on n300 on device 0 only (which is what issue #3934 is for). Looking into getting things working/testing for the remote device now, but will be a separate PR.

File: tt_metal/llrt/tt_cluster.cpp (hunk 1 above)
Comment (TT-billteng) [reply tt_metal/llrt/tt_cluster.cpp:351]: I'm thinking we should seriously consider explicit typed coordinates now (virtual, physical, ... )

**Review by TT-billteng — APPROVED** [review]

File: tt_metal/llrt/tt_cluster.cpp (hunk 1 above)
Comment (aliuTT) [reply tt_metal/llrt/tt_cluster.cpp:351]: We sort of have the abstraction layer called `metal_soc_desc`. we can probably make `get_soc_desc` unaccessible outside of the constructor of metal soc desc.

**Review by pgkeller — APPROVED** [review]

**Review by DrJessop — APPROVED** [review]

**Comment by tt-dma** [comment]
Passing CI: https://github.com/tenstorrent-metal/tt-metal/actions/runs/7281391145. Will merge

**Pull request merged**